fuel-vm/src/interpreter/blob.rs

```diff
@@ -0,0 +1,94 @@
+use fuel_asm::{
+    RegisterId,
+    Word,
+};
+use fuel_storage::StorageSize;
+use fuel_tx::PanicReason;
+use fuel_types::BlobId;
+
+use crate::{
+    error::IoResult,
+    prelude::*,
+    storage::{
+        BlobData,
+        InterpreterStorage,
+    },
+};
+
+use super::{
+    internal::inc_pc,
+    split_registers,
+    GetRegMut,
+    Interpreter,
+    Memory,
+    SystemRegisters,
+    WriteRegKey,
+};
+
+impl<M, S, Tx, Ecal> Interpreter<M, S, Tx, Ecal>
+where
+    M: Memory,
+    S: InterpreterStorage,
+    <S as InterpreterStorage>::DataError: From<S::DataError>,
+{
+    pub(crate) fn blob_size(
+        &mut self,
+        dst: RegisterId,
+        blob_id_ptr: Word,
+    ) -> IoResult<(), S::DataError> {
+        let gas_cost = self
+            .interpreter_params
+            .gas_costs
+            .bsiz()
+            .map_err(PanicReason::from)?;
+        self.gas_charge(gas_cost.base())?;
+
+        let blob_id = BlobId::from(self.memory.as_ref().read_bytes(blob_id_ptr)?);
+
+        let size = <S as StorageSize<BlobData>>::size_of_value(&self.storage, &blob_id)
+            .map_err(RuntimeError::Storage)?
+            .ok_or(PanicReason::BlobNotFound)?;
+
+        self.dependent_gas_charge_without_base(gas_cost, size as Word)?;
+        let (SystemRegisters { pc, .. }, mut w) = split_registers(&mut self.registers);
+        let result = &mut w[WriteRegKey::try_from(dst)?];
+        *result = size as Word;
+        Ok(inc_pc(pc)?)
+    }
+
+    pub(crate) fn blob_load_data(
+        &mut self,
+        dst_ptr: Word,
+        blob_id_ptr: Word,
+        blob_offset: Word,
+        len: Word,
+    ) -> IoResult<(), S::DataError> {
+        let gas_cost = self
+            .interpreter_params
+            .gas_costs
+            .bldd()
+            .map_err(PanicReason::from)?;
+        self.dependent_gas_charge(gas_cost, len)?;
+
+        let blob_offset: usize = blob_offset
+            .try_into()
+            .map_err(|_| PanicReason::MemoryOverflow)?;
+
+        let blob_id = BlobId::from(self.memory.as_ref().read_bytes(blob_id_ptr)?);
+        let owner = self.ownership_registers();
+        let dst = self.memory.as_mut().write(owner, dst_ptr, len)?;
+
+        let blob = <S as StorageInspect<BlobData>>::get(&self.storage, &blob_id)
+            .map_err(RuntimeError::Storage)?
+            .ok_or(PanicReason::BlobNotFound)?;
+        let blob = blob.as_ref().as_ref();
+
+        let end = blob_offset.saturating_add(dst.len()).min(blob.len());
```

Comment:
The `blob_offset` could be `u32::MAX` that could cause different result for 32 and 64 bit platforms. We need to check that ensures that `blob_offset` less than blob size. In the code copy case we check that it is less than max contract size(What also could be incorrect, maybe we want to compare it with the size of the contract).

Reply:
Hmm, even if saturation occurs it will be handled by `min` later. But maybe it is still better to add this check to avoid potential problems in the future

Reply:
Hmm, the logic here looks like the logic inside of the `copy_from_slice_zero_fill`. Maybe we can use it here instead(and the same for code copy)?

Reply:
Reading the spec, there's no need for the offset to be less than the contract size. Fixes done to other opcodes (CCP, LDC) are incorrect (against the spec). The spec doesn't mention panicing even If the offset is larger than the size. Is the spec wrong here? It would feel weird to special-case fitting into `u32`.

Reply:
I fixed all those in 34107946.